content/posts/numpy/generator-rng/index.md

```diff
@@ -0,0 +1,304 @@
+np.random.seed(123456)
+sample = np.random.random(1_024)
+```
+
+Aoutch! Two lines of codes and both are wrong. This is a simple snippet to
+generate 1024 points using NumPy's random number generator that is very common
+in codes and tutorials online. Sadly/surprisingly even on famous blogs.
+
+There are 2 problems I want to talk about:
+
+1. [global state](#global-id),
+2. [seed](#seed-id).
+
+## Goodbye global state {#global-id}
+
+Behind `np.random.random` a number generator is used. This generator has an
+internal state which is _global_. This state is shared and as new points/samples
+are requested, this state is used and updated. The state can be set with a
+_seed_ value.
+
+As with any global variable, it means that it can be changed by other
+subprocesses, threads, other libraries, etc. Take the following code. It uses
+4 threads to call `np.random.random` 10 times and then prints the last value.
+We execute this 5 times to see how the value changes.
+
+```python
+# the following is wrong and only exists for demonstration purposes!
+import concurrent.futures
+
+for _ in range(5):
+    np.random.seed(123456)  # fix the state of the global generator
+
+    sample = []
+    with concurrent.futures.ThreadPoolExecutor(4) as executor:
+        futures = [executor.submit(np.random.random, 1) for _ in range(10)]
+        for future in concurrent.futures.as_completed(futures):
+            sample.append(future.result())
+
+    print(sample[-1])
+
+# [0.33622174]
+# [0.5430262]
+# [0.96671784]
+# [0.37674972]
+# [0.96671784]
+```
+
+As you can see, although we started with the same state, this program is not
+deterministic anymore. This is what we mean by thread safety.
+
+NumPy came up with a solution: `np.random.RandomState`. The idea is,
+knowing random number generators have a state, let's define them as
+objects. This way, each object can have it's internal state instead of relying
+on a global one. This API is more explicit and prevents any "contamination"
+of state. See this example.
+
+```python
+# the following is wrong and only exists for demonstration purposes!
+seed = 123456
+
+# reference value
+rng = np.random.RandomState(seed)
+print(rng.random(6)[-1])
+
+np.random.seed(seed)
+rng = np.random.RandomState(seed)
+
+for _ in range(5):
+    np.random.random(1)
+    rng.random(1)
+
+print(rng.random(1)[0])
+
+# 0.33622174433445307
+# 0.33622174433445307
+```
+
+It sets the global seed and defines a `rng` using the object based API.
+
+> RNG stands for Random Number Generator. This abbreviation is commonly used.
+
+Then `np.random.random` and `np.random.RandomState` calls 5 times. The global
+state is getting updated, but it does not impact our `rng` stream of numbers.
+
+A lot of packages already migrated their code to move away from the global
+generator in favour of an object one. In this case, setting the global state
+would have absolutely no effect and you will end up in a broken state.
```

Comment:
In generally seeding isn't needed or particulary beneficial for existing `RandomState` objects.  Might as well create a new instance with the seed value than calling `seed`.